src/python/pants/engine/legacy/graph.py

```diff
@@ -313,21 +314,54 @@ def __hash__(self):
     return hash(self.address)
 
 
+class TransitiveHydratedTarget(datatype('TransitiveHydratedTarget', ['root', 'dependencies'])):
+  """A recursive structure wrapping a HydratedTarget root and TransitiveHydratedTarget deps."""
+
+
 class TransitiveHydratedTargets(Collection.of(HydratedTarget)):
-  """A transitive set of HydratedTarget objects."""
+  """A transitive set of HydratedTarget objects, flattened and de-duped."""
 
 
 class HydratedTargets(Collection.of(HydratedTarget)):
   """An intransitive set of HydratedTarget objects."""
 
 
-@rule(TransitiveHydratedTargets, [SelectTransitive(HydratedTarget,
-                                                   BuildFileAddresses,
-                                                   field_types=(Address,),
-                                                   field='addresses')])
-def transitive_hydrated_targets(targets):
-  """Recursively requests HydratedTarget instances, which will result in an eager, transitive graph walk."""
-  return TransitiveHydratedTargets(targets)
+@rule(TransitiveHydratedTargets, [SelectDependencies(TransitiveHydratedTarget,
+                                                     BuildFileAddresses,
+                                                     field_types=(Address,),
+                                                     field='addresses')])
+def transitive_hydrated_targets(transitive_hydrated_targets):
+  """Kicks off recursion on expansion of TransitiveHydratedTarget objects.
+
+  The TransitiveHydratedTarget struct represents a structure-shared graph, which we walk
+  and flatten here. The engine memoizes the computation of TransitiveHydratedTarget, so
+  when multiple TransitiveHydratedTargets objects are being constructed for multiple
+  roots, their structure will be shared.
+
+  TODO: Further work would be to create exactly one `Collection.of(Spec)` root that also
+  dedupes.
+  """
+  result = []
```

Comment:
Either document the ordering guarantee here, or use a set

Reply:
Good point! Thanks.